ethereum/eip712/eip712_fuzzer_test.go

```diff
@@ -0,0 +1,140 @@
+package eip712_test
+
+import (
+	"fmt"
+	"strings"
+
+	rand "github.com/tendermint/tendermint/libs/rand"
+
+	"github.com/evmos/evmos/v11/ethereum/eip712"
+	"github.com/tidwall/gjson"
+	"github.com/tidwall/sjson"
+)
+
+// TestRandomPayloadFlattening generates many random payloads with different JSON values to ensure
+// that Flattening works across all inputs.
+// Note that this is a fuzz test, although it doesn't use Go's Fuzz testing suite, since there are
+// variable input sizes, types, and fields. While it may be possible to translate a single input into
+// a JSON object, it would require difficult parsing, and ultimately approximates our randomized unit
+// tests as they are.
+func (suite *EIP712TestSuite) TestRandomPayloadFlattening() {
+	// Re-seed rand generator
+	rand.Seed(rand.Int64())
+
+	numTestObjects := 15
+	for i := 0; i < numTestObjects; i++ {
+		suite.Run(fmt.Sprintf("Flatten%d", i), func() {
+			payload := suite.generateRandomPayload(i)
+
+			flattened, numMessages, err := eip712.FlattenPayloadMessages(payload)
+
+			suite.Require().NoError(err)
+			suite.Require().Equal(numMessages, i)
+
+			suite.verifyPayloadAgainstFlattened(payload, flattened)
+		})
+	}
+}
+
+// generateRandomPayload creates a random payload of the desired format, with random sub-objects.
+func (suite *EIP712TestSuite) generateRandomPayload(numMessages int) gjson.Result {
+	payload := suite.createRandomJSONObject().Raw
+	msgs := make([]gjson.Result, numMessages)
+
+	for i := 0; i < numMessages; i++ {
+		m := suite.createRandomJSONObject()
+		msgs[i] = m
+	}
+
+	payload, err := sjson.Set(payload, "msgs", msgs)
+	suite.Require().NoError(err)
+
+	return gjson.Parse(payload)
+}
+
+// createRandomJSONObject creates a JSON object with random fields.
+func (suite *EIP712TestSuite) createRandomJSONObject() gjson.Result {
```

Comment:
I wrote more context on this in the original PR body and the Notion spec, but to give an overview:

The current `TypedData` implementation reads the `sdk.Msg` types in Golang and runs through the fields to try to guess which ones are in the payload. This is difficult, however, as we don't know which fields are actually present in the payload—which leads to errors.

I decided instead that we should read the JSON payload itself to build the types, since this avoids any ambiguity—by dealing directly with the final representation, we can better understand which fields are used, since they are either present in the final JSON, or not.

I had a version of this PR without using `gjson`, but rather parsing each JSON object using Go's `map[string]interface{}` type. Since we're dealing with JSON format, however, I decided we could have much cleaner code if we used a library built to handle it, as the syntax and manipulation would be much easier.

We use `gjson` ([Getting JSON](https://github.com/tidwall/gjson)) and `sjson` ([Setting JSON](https://github.com/tidwall/sjson)), which are both well-used and well-tested.